### PR TITLE
Add subscribe_options class, and convert qos and retain_handling to enums which automatically convert

### PR DESCRIPTION
**Pull request by jonesmz**
Replaces https://github.com/redboltz/mqtt_cpp/pull/379

### CODE REVIEW

**Comment by jonesmz** [comment]
Note that almost all of the static asserts that are added by this PR can be removed if a similar treatment is given to the success_code, reason_code, and other values.

File: include/mqtt/subscribe_options.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    char const * const str[] = {
+        "at_most_once",
+        "at_least_once",
+        "exactly_once"
+    };
+    if (static_cast<std::uint8_t>(v) < sizeof(str)) return str[static_cast<std::uint8_t>(v)];
+    return "invalid_qos";
+}
+
+template<typename STREAM_T>
```
Comment (redboltz) [inline include/mqtt/subscribe_options.hpp]:
```suggestion
template<typename Stream>
```

See https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules

Comment (jonesmz) [reply include/mqtt/subscribe_options.hpp]: Fixed

**Comment by redboltz** [comment]
> Note that almost all of the static asserts that are added by this PR can be removed if a similar treatment is given to the success_code, reason_code, and other values

Do you mean static_cast ?

**Comment by jonesmz** [comment]
Yes.

I meant static_cast.

Sorry for the confusion.

**Comment by redboltz** [comment]
Thank you!
If you create a PR for reason_code and others to be `enum class`, I would appreciate.

**Comment by jonesmz** [comment]
Thanks for merging this.

It's already helped me catch a bug in my code.

```
pMqttClient->async_publish(boost::asio::buffer(topic),
                                           {},
                                           topic, // lifetime
                                           false,
                                           redboltz::qos::exactly_once,
                                           { redboltz::v5::property::correlation_data(redboltz::allocate_buffer(requestID)) });
```

Notice the QOS and Retain parameters are out of order.

Before this change, the bool argument was being implicitly converted to std::uint8_t, and the std::uint8_t was being converted to bool.

Now that you've merged my patch, that call fails to compile, as it should have all along.

**Pull request merged**